tempesta_fw/sched/tfw_sched_hash.c

```diff
@@ -114,42 +229,73 @@ static int
 tfw_sched_hash_del_srv(TfwServer *srv)
 {
 	int ret = 0;
-	int i;
-
-	spin_lock_bh(&servers_write_lock);
-	i = find_srv_idx(srv);
-	if (i < 0) {
-		ERR("Can't delete the server from the scheduler - not found\n");
-		ret = -ENOENT;
-	} else {
-		servers[i] = servers[servers_n - 1];
-		--servers_n;
-		servers[servers_n] = NULL;
+	TfwSrvHash *curr;
+	TfwSrvHashList *new_list, *old_list;
+	size_t new_list_size, old_list_size;
+
+	spin_lock_bh(&srv_hashes_update_lock);
+
+	old_list = tfw_srv_hash_list;
+	old_list_size = TFW_SRV_HASH_LIST_SIZE(old_list->n);
+	new_list_size = TFW_SRV_HASH_LIST_SIZE(old_list->n - 1);
```

Comment:
Actually there is no need to shrink the list - ever 64K servers is just about 1MB of memory.